### ISSUE TITLE
[BUG] TS2339: Property 'on' does not exist on type.

### ISSUE THREAD

**Issue by sebastien-savalle** [cap-js/cds-types#318]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

We observed a weird behavior when migrating to `cds` 8 and `cds-types` 0.7.0. 

Custom handler cannot be registered properly. A typescript error is thrown. 

`error TS2339: Property 'on' does not exist on type 'AdminService'.`

```
import { ApplicationService } from '@sap/cds';

export default class AdminService extends ApplicationService {
  init() {
    this.on('health', () => {
      return { status: 'UP' };
    });
  }
}
```

But it works when this is cast to ApplicationService.

```
import { ApplicationService } from '@sap/cds';

export default class AdminService extends ApplicationService {
  init() {
    (this as ApplicationService).on('health', () => {
      return { status: 'UP' };
    });
  }
}
```

### References

https://pages.github.tools.sap/cap/docs/tools/cds-typer#typer-top-level-imports

### Versions

| backend      | https://github.com/<your/repo>                                                                                          |
| ---------------------- | ----------------------------------------------------------------------------------------------------------------------- |
| @cap-js/asyncapi       | 1.0.2                                                                                                                   |
| @cap-js/audit-logging  | 0.8.1                                                                                                                   |
| @cap-js/cds-types      | 0.7.0                                                                                                                   |
| @cap-js/db-service     | 1.14.1                                                                                                                  |
| @cap-js/hana           | 1.4.1                                                                                                                   |
| @cap-js/openapi        | 1.0.7                                                                                                                   |
| @cap-js/sqlite         | 1.7.6                                                                                                                   |
| @cap-js/telemetry      | 1.0.1                                                                                                                   |
| @sap/cds               | 8.4.0                                                                                                                   |
| @sap/cds-compiler      | 5.4.0                                                                                                                   |
| @sap/cds-dk            | 8.4.0                                                                                                                   |
| @sap/cds-dk (global)   | 8.4.0                                                                                                                   |
| @sap/cds-fiori         | 1.2.7                                                                                                                   |
| @sap/cds-foss          | 5.0.1                                                                                                                   |
| @sap/cds-mtxs          | 2.3.0                                                                                                                   |
| @sap/eslint-plugin-cds | 3.0.4                                                                                                                   |
| Node.js                | v20.15.1                                                                                                                |

**Comment by daogrady** [comment]
Hi Sébastien,

thanks for reporting this issue!
This looks a bit like TS is trying to infer the type from the implementation files, which would imply you don't have cds-types installed in your project.
Please note that since cds8, `@cap-js/cds-types` has to be installed explicitly and is no longer pulled in from `@sap/cds`.
Did you add `@cap-js/cds-types` as devDependency to your project and install it?

Best,
Daniel

**Comment by daogrady** [comment]
Actually, I am pretty sure this is a duplicate of https://github.com/cap-js/cds-types/issues/251

If you do have cds-types installed, can you please run `npm rebuild` in your project and see, if that fixes the problem?

**Comment by sebastien-savalle** [comment]
Hi Daniel,

Yes, we do have cds-types installed as you can see in the summary. 
We use `pnpm`, so this might be similar to the issue you mentioned.

But, this works well with version `0.6.5`.

**Comment by daogrady** [comment]
Hi Sébastien,

thanks for checking! Please also check out [my other suggestion](https://github.com/cap-js/cds-types/issues/318#issuecomment-2467590435).

Best,
Daniel

**Comment by daogrady** [comment]
Hi Sébastien,

can you please check out @hakimio's suggestion and see if that solves the problem for you?
If it does, could you please also elaborate a bit on your setup? Do you use a monorepo? If so, do the submodules of that monorepo have a dependency on cds-types?

Best,
Daniel

**Comment by hakimio** [comment]
@sebastien-savalle how does your `tsconfig` look like? Does the `types` config work if you set `moduleResolution` to `NodeNext`?

**Comment by sebastien-savalle** [comment]
No, this does not work even with `moduleResolution` set to `NodeNext`

And `pnpm rebuild` is sometimes helpful when everything looks broken.

tsconfig:

```
{
  "compilerOptions": {
    /* Basic Options */
    "target": "es2019",
    "module": "nodenext",
    "outDir": "./gen",
    "rootDir": "./",
    "baseUrl": ".",
    "moduleResolution": "nodenext",
    "skipLibCheck": true,
    "preserveConstEnums": true,
    "sourceMap": true,
    "allowJs": true,
    "strict": true,
    "strictNullChecks": true,
    "strictPropertyInitialization": false,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    /* Experimental Options */
    "noImplicitAny": false,
    "emitDecoratorMetadata": true, /* Needed for the decorators in the Entities.ts and FunctionHandler.ts. Enables experimental support for emitting type metadata for decorators. */
    "experimentalDecorators": true, /* Enables experimental support for ES7 decorators. */
    "types": [
      "node",
      "jest",
      "express"
    ],
    "typeRoots": [
      "./types",
      "./node_modules/@types",
      "./node_modules/@cap-js/cds-types"
    ]
  },
  "include": [
    "../.eslintrc.json",
    "./srv/**/*.ts",
    "./types/**/**.d.ts",
    "./test/**/*.ts"
  ]
}
```
